src/r2b2/minerva.py

```diff
@@ -0,0 +1,218 @@
+"""Minerva audit module."""
+import math
+from typing import List
+
+import click
+from scipy.stats import binom
+
+from r2b2.audit import Audit
+from r2b2.contest import Contest
+
+
+class Minerva(Audit):
+    """Minerva audit implementation.
+
+    A Minerva audit is a type of risk-limiting audit that accounts for round-by-round auditor
+    decisions. For a given sample size (in the context of a round schedule), the audit software
+    calculates a minimum number of votes for the reported winner that must be found in the sample
+    to stop the audit and confirm the reported outcome.
+
+    Attributes:
+        alpha (float): Risk limit. Alpha represents the chance that, given an incorrectly called
+        election, the audit will fail to force a full recount.
+        max_fraction_to_draw (float): The maximum number of ballots the auditors are willing to draw
+        as a fraction of the ballots in the contest.
+        rounds (List[int]): Cumulative round schedule.
+        min_winner_ballots (List[int]): Stopping sizes (or kmins) respective to the round schedule.
+        contest (Contest): Contest to be audited.
+    """
+
+    def __init__(self, alpha: float, max_fraction_to_draw: float, contest: Contest):
+        """Initialize a Minerva audit."""
+        super().__init__(alpha, 0.0, max_fraction_to_draw, True, contest)
+        self.min_sample_size = self.get_min_sample_size()
+        self.rounds = []
+        self.min_winner_ballots = []
+
+    def get_min_sample_size(self, min_sprob: float = 10 ** (-6)):
+        """Computes the minimum sample size that has a stopping size (kmin). Here we find a
+        practical minimum instead of the theoretical minimum (BRAVO's minimum) to avoid
+        floating-point imprecisions in the later convolution process.
+
+        Args:
+            min_sprob (float): Round sizes with below min_sprob stopping probability are excluded.
+        Returns:
+            int: The minimum sample size of the audit, adherent to the min_sprob.
+        """
+
+        # p0 is not .5 for contests with odd total ballots.
+        p0 = (self.contest.contest_ballots // 2) / self.contest.contest_ballots
+        p1 = self.contest.winner_prop
+        max_sample_size = math.ceil(self.contest.contest_ballots * self.max_fraction_to_draw)
+
+        # There may not be a point n' such that all n >= n' are acceptable round sizes and all
+        # n < n' are unacceptable round sizes.
+        for n in range(1, max_sample_size):
+            num_dist = binom.pmf(range(n+1), n, p1)
+            denom_dist = binom.pmf(range(n+1), n, p0)
+            if self.satisfactory_sample_size(n // 2, n+1, min_sprob, num_dist, denom_dist):
+                return n
```

Comment:
One idea I had a while ago (which I totally forgot to talk to you about) for making the search here faster in all the audits is the idea of a max-min sample size. The method in brla was getting really expensive for larger contests, so I ended up running a script that just created a bunch of brla audits for a range of risk limits and contest sizes. I ran it and just recorded all the min sample sizes and found that they never really went above 20 or 30. 

Essentially, what if we could hardcode the search max instead of having it be the max fraction to draw? 

I was playing around with the idea probably close to 2 months ago, so there could be a reason that it doesn't work and I've forgotten that. I also haven't given it much thought outside the context of BRLA, but I'm interested to hear what you think. 

My intuition tells me won't be able to find a hardcode-able max for athena/minerva/metis since we need to consider the margin, but what do you think about investigating the idea of a *'more informed'* max for the search?

It might not be worth the effort, but now that we have access to the Nimbus cluster we have enough compute power to generate **a lot** of min sample sizes to analyze.

Reply:
I agree with your intuition that this would be hard for Athena-class audits. Sadly, we cannot even use the inverse hypergeometric function since the round size varies (and, additionally, we're inspecting a ratio). I think to improve this approach would require something analytical (it was originally implemented as the absolute BRAVO minimum, but for narrower margins such a sample size may have a stopping probability of, say, 10^-40), or perhaps a binary search. But a binary search may be troublesome in its own way since a round size n may be satisfactory where a round size of n+1 isn't (as we saw in the map data).

In terms of changing the max to a more informed value, perhaps I'm misunderstanding your point, but I'm not sure it would help that much since (1) we absolutely need to find a minimum sample size and (2) we will iterate up to the acceptable value (and not the maximum) anyway; if the minimum sample is 100 then we stop the loop after 100 irrespective of the maximum.

But yes, this algorithm can indeed become incredibly computationally expensive, and it might be worthwhile to explore a "binary approximation" where we see if the behavior holds in the neighborhood of the proposed minimum. I'll give this some more thought. Thanks!

Reply:
After talking about this (briefly) with Poorvi, I think I can refine what I meant, which in retrospect looks like scattered ramblings. I realize why changing the max is not as helpful for Athena, since the search is linear. It improved the BRLA significantly, but only because it can easily perform a binary search. The concept I was trying to get at though is reducing the search space. 

For example, perhaps we learn how to identify a case where the linea search becomes expensive (e.g. margin < tolerance) and for these cases, we perform the search from some lower bound instead. Essentially, if we can identify cases where the minimum is always above n, then we start the search from n instead of 1.

Another idea is to generate a table of minimum sample sizes and always start from a data point that will less than or equal to the desired minimum sample size (if that makes sense).